README.md
**BotTweeter**

![Image](https://i.imgur.com/MY8Hs0K.png)

DISCLAIMER
-----------------------------------------------------------------------------------
BT was made to run on APIv1, I'm pretty sure (but not certain) that it works on
APIv2 natively though. If you're getting authentication errors after verifying
that the tokens are correct, the API version used is most likely the issue.
-----------------------------------------------------------------------------------

INSTALLATION
-----------------------------------------------------------------------------------
-Install pip if you haven't already

-Open setup.py

-The script will install all required libraries

-Enter your authentication info

-If some info is incorrect, delete auth.data and run the setup again

-Once setup.py is finished you can use BotTweeter by double-clicking BotTweeter.py
-----------------------------------------------------------------------------------


Changelog.txt
1.1.0
-Ability to reply to tweets
-Ability to post and reply with media

1.2.0
-Drafts
-Setup wizard

1.3.0
-Now using up to date tweepy
-Errors are now logged in crashlog.txt
-Character limit check
-Secure version of BT (all authentication info is encrypted)
-WARNING: You will not be able to use BT on another machine without setting it up again

1.3.1
-Fixed setup package installer bug
-Cleaned up code, now it is a lot more readable and efficient (WIP, I might optimize it further)

1.3.2
-Fixed a slight issue with default filename entry
-Fixed a bug with selecting media to be uploaded
-Cleaned up the code further
-Error logs are now a lot more consistent
-Fixed bug with replying with media

Todo:
Ability to select media from outside the BT folder
Ability to quote tweet without pasting link
Tweet multiple images in single tweet

drafts.py
import sys
import tkinter as tk
import tkinter.ttk as ttk
import drafts_support
import BotTweeter
from BotTweeter import vp_start_gui
import json
import os

def vp_start_gui():
    global val, w, root
    root = tk.Tk()
    top = Toplevel1 (root)
    drafts_support.init(root, top)
    root.mainloop()

w = None
def create_Toplevel1(rt, *args, **kwargs):
    global w, w_win, root
    root = rt
    w = tk.Toplevel (root)
    top = Toplevel1 (w)
    drafts_support.init(w, top, *args, **kwargs)
    return (w, top)

def destroy_Toplevel1():
    global w
    w.destroy()
    w = None

class Toplevel1:
    def __init__(self, top=None):
        _bgcolor = '#d9d9d9'
        _fgcolor = '#000000'
        _compcolor = '#d9d9d9'
        _ana1color = '#d9d9d9'
        _ana2color = '#ececec'

        top.geometry("600x173+765+302")
        top.minsize(120, 1)
        top.maxsize(1924, 1061)
        top.resizable(1,  1)
        top.title("New Toplevel")
        top.configure(background="#d9d9d9")
        
        f = open('drafts.paul.town', 'r')
        global currentpage
        currentpage = 1
        lines = 0
        for line in f:
            if line != "\n":
                lines += 1
        f.close()

        global Dtweet
        global DTID
        global Dmedia
        Dtweet = str
        DTID = str
        Dmedia = str

        self.Label1 = tk.Label(top)
        self.Label1.place(relx=0.4, rely=0.675, height=41, width=114)
        self.Label1.configure(background="#d9d9d9")
        self.Label1.configure(disabledforeground="#a3a3a3")
        self.Label1.configure(foreground="#000000")
        self.Label1.configure(text='Page: '+str(currentpage))
        self.Label1.configure(font="-family {Arial} -size 16 -weight bold")
        self.Label1.configure(highlightcolor="black")

        def display_draft():
            global currentpage
            global cdraft
            with open('drafts.paul.town') as rawdrafts:
                draftlist = list(rawdrafts)
            print(draftlist[currentpage-1])
            cdraft = json.loads(draftlist[currentpage-1])
            self.Button1.config(text=cdraft.get('tweet'))

        def next():
            global currentpage
            global cdraft
            if currentpage < lines:
                currentpage += 1
                print(currentpage)
                self.Label1.config(text='Page: '+str(currentpage))
                self.Button2['state'] = 'normal'
                display_draft()
            if currentpage == lines:
                self.Button3['state'] = 'disabled'
                display_draft()
            
        def prev():
            global currentpage
            if currentpage >= 2:
                currentpage -= 1
                self.Button3['state'] = 'normal'
                self.Label1.config(text='Page: '+str(currentpage))
                if currentpage == 1:
                    self.Button2['state'] = 'disabled'
                display_draft()
                
        def select_draft():
            global cdraft
            global Dtweet
            global DTID
            global Dmedia
            Dtweet = cdraft.get('tweet')
            if 'TID' in cdraft:
                DTID = cdraft.get('TID')
            if 'media' in cdraft:
                Dmedia = cdraft.get('media')

            #BotTweeter.Toplevel1().__init__(self).draftmode()
            with open('temp.dump', 'w') as f:
                json.dump(cdraft, f)
            os.execv(sys.executable, ['python BotTweeter.py'] + sys.argv)

        self.Button1 = tk.Button(top)
        self.Button1.place(relx=0.017, rely=0.04, height=94, width=577)
        self.Button1.configure(activebackground="#ececec")
        self.Button1.configure(activeforeground="#000000")
        self.Button1.configure(background="#d9d9d9")
        self.Button1.configure(disabledforeground="#a3a3a3")
        self.Button1.configure(foreground="#000000")
        self.Button1.configure(highlightbackground="#d9d9d9")
        self.Button1.configure(highlightcolor="black")
        self.Button1.configure(pady="0")
        self.Button1.configure(text='Button')
        self.Button1.configure(command=select_draft)

        self.Button2 = tk.Button(top)
        self.Button2.place(relx=0.017, rely=0.636, height=54, width=117)
        self.Button2.configure(activebackground="#ececec")
        self.Button2.configure(activeforeground="#000000")
        self.Button2.configure(background="#d9d9d9")
        self.Button2.configure(disabledforeground="#a3a3a3")
        self.Button2.configure(foreground="#000000")
        self.Button2.configure(highlightbackground="#d9d9d9")
        self.Button2.configure(highlightcolor="black")
        self.Button2.configure(pady="0")
        self.Button2.configure(text='Previous')
        self.Button2['state'] = 'disabled'
        self.Button2.configure(command=prev)
        self.Button2.configure(font="-family {Arial} -size 18 -weight bold")
        self.Button2.configure(highlightcolor="black")

        self.Button3 = tk.Button(top)
        self.Button3.place(relx=0.783, rely=0.636, height=54, width=117)
        self.Button3.configure(activebackground="#ececec")
        self.Button3.configure(activeforeground="#000000")
        self.Button3.configure(background="#d9d9d9")
        self.Button3.configure(disabledforeground="#a3a3a3")
        self.Button3.configure(foreground="#000000")
        self.Button3.configure(highlightbackground="#d9d9d9")
        self.Button3.configure(highlightcolor="black")
        self.Button3.configure(pady="0")
        self.Button3.configure(text='Next')
        self.Button3.configure(command=next)
        self.Button3.configure(font="-family {Arial} -size 18 -weight bold")
        self.Button3.configure(highlightcolor="black")

        if lines == 1:
            self.Button2['state'] = 'disabled'
            self.Button3['state'] = 'disabled'

        display_draft()

if __name__ == '__main__':
    vp_start_gui()







BotTweeter.py
import sys
import traceback
import tweepy
import os
from tkinter import messagebox as mbox
import os.path
import tkinter as tk
import tkinter.ttk as ttk
import BT_support
import drafts
import subprocess
from Cryptodome.Cipher import AES
uuid = str(subprocess.check_output('wmic csproduct get uuid'), 'utf-8').split('\n')[1].strip().split('-')
ekey = uuid[0]+uuid[2]+uuid[1]

with open('auth.data', 'rb') as authdata:
    authlist = authdata.read().split('\n'.encode('utf-8'))
decrypt = AES.new(ekey.encode("utf8"), AES.MODE_CFB, 'BotTweeterMadlad'.encode("utf8"))
auth = decrypt.decrypt(authlist[0]).decode('utf-8')
authsecret = decrypt.decrypt(authlist[1]).decode('utf-8')
key = decrypt.decrypt(authlist[2]).decode('utf-8')
keysecret = decrypt.decrypt(authlist[3]).decode('utf-8')

#####AUTHENTICATION#####
auth = tweepy.OAuthHandler(auth, authsecret) #ENTER INFO HERE
auth.set_access_token(key, keysecret) #ENTER INFO HERE

api = tweepy.API(auth)

try:
    api.verify_credentials()
    print("Authentication confirm")
    authed = True
except:
    print("Error during authentication")

def vp_start_gui():
    global val, w, root
    root = tk.Tk()
    BT_support.set_Tk_var()
    top = Toplevel1 (root)
    BT_support.init(root, top)
    root.mainloop()


class Toplevel1:
    def __init__(self, top=None):

        def crashlog(exc):
            with open('crashlog.txt', 'w') as f:
                f.write(str(exc))
            mbox.showinfo('Error','''Something went wrong! Please contact the developer and send them the crashlog.txt''', icon='error')


        _bgcolor = '#d9d9d9'

        top.geometry("575x350") #+713+216"
        top.minsize(120, 1)
        top.maxsize(1924, 1061)
        top.resizable(1,  1)
        top.title("BotTweeter 1.3.2")
        top.configure(background=_bgcolor)
        top.configure(highlightbackground=_bgcolor)
        top.configure(highlightcolor="black")

        self.title = tk.Label(top)
        self.title.place(relx=0.299, rely=0.029, height=32, width=220) #relx = .299 when subversion is included (x.y.z), otherwise .313
        self.title.configure(text='BotTweeter 1.3.2', background=_bgcolor, font="-size 20 -weight bold")

        self.sig = tk.Label(top)
        self.sig.place(relx=0.417, rely=0.114, height=32, width=71)
        self.sig.configure(text='by madlad', background=_bgcolor, font=" size 12")

        self.captionLabel = tk.Label(top)
        self.captionLabel.place(relx=0.0, rely=0.149, height=18, width=124)
        self.captionLabel.configure(text='Tweet caption', background=_bgcolor, font="-size 13")

        cb1 = tk.IntVar()
        cb2 = tk.IntVar()

        self.replyID = tk.Entry(top)
        self.replyID.place(relx=0.017, rely=0.714, height=20, relwidth=0.337)
        self.replyID.configure(font="-family {Courier New} -size 11")


        self.replyIDLabel = tk.Label(top)
        self.replyIDLabel.place(relx=0.017, rely=0.629, height=18, width=74)
        self.replyIDLabel.configure(text='Tweet ID', background=_bgcolor, font="-size 13")

        self.filenameLabel = tk.Label(top)
        self.filenameLabel.place(relx=0.017, rely=0.8, height=18, width=124)
        self.filenameLabel.configure(text='Media filename', background=_bgcolor, font="-size 13")

        self.filename = tk.Entry(top)
        self.filename.place(relx=0.017, rely=0.886, height=20, relwidth=0.337)
        self.filename.configure(font="-family {Courier New} -size 11")
        self.filename.insert(0,'default.jpg')

        self.caption = tk.Text(top)
        self.caption.place(relx=0.017, rely=0.229, relheight=0.269, relwidth=0.877)
        self.caption.configure(font="-family {Segoe UI} -size 11", wrap="word")

        def endis():
            if cb1.get() == 0:
                self.replyID['state'] = 'disabled'
            else:
                self.replyID['state'] = 'normal'
            if cb2.get() == 0:
                self.filename['state'] = 'disabled'
            else:
                self.filename['state'] = 'normal'

        self.Checkbutton2 = tk.Checkbutton(top)
        self.Checkbutton2.place(relx=0.174, rely=0.514, relheight=0.086, relwidth=0.245)
        self.Checkbutton2.configure(background=_bgcolor, command=endis, variable=cb2, text='Contains media', justify='left', font="-size 13")

        self.Checkbutton1 = tk.Checkbutton(top)
        self.Checkbutton1.place(relx=0.017, rely=0.514, relheight=0.086, relwidth=0.123)
        self.Checkbutton1.configure(font="-size 13", background=_bgcolor, variable=cb1, command=endis, text='Reply', justify='left')
        
        def read_drafts():
            if os.stat('drafts.paul.town').st_size != 0:
                try: drafts.vp_start_gui()
                except Exception as e:
                    crashlog(exc = traceback.format_exc())
                    exit()
            else:
                mbox.showinfo('Error','You have no drafts')
        def save_drafts():
            with open('drafts.paul.town', 'a') as f:
                if cb1.get() == 1:
                    if cb2.get() == 1:
                        if self.replyID.get() != '':
                            if self.filename.get() != '':
                                if os.path.isfile(self.filename.get()) == True:
                                    try:
                                        draft = {'tweet':self.caption.get('1.0','end-1c'), 'TID':self.replyID.get(), 'media':self.filename.get()}
                                        json.dump(draft, f)
                                        f.write('\n')
                                        print('Draft saved!')
                                        mbox.showinfo('Success!','Draft has been saved')
                                    except Exception as e:
                                        crashlog(exc = traceback.format_exc())
                                        exit()
                                else:
                                    mbox.showinfo('Error','''No such file exists in this script's folder''')
                            else:
                                mbox.showinfo('Error','Please specify a media file')
                        else:
                            mbox.showinfo('Error','There is no tweet ID')
                    else:
                        if self.caption.get('1.0','end-1c') != '':
                            if self.replyID.get() != '':
                                draft = {'tweet':self.caption.get('1.0','end-1c'), 'TID':self.replyID.get()}
                                json.dump(draft, f)
                                f.write('\n')
                                print('Draft saved!')
                                mbox.showinfo('Success!','Draft has been saved')
                            else:
                                mbox.showinfo('Error','There is no tweet ID')
                        else:
                            mbox.showinfo('Error','Your tweet is empty')

                else:
                    if cb2.get() == 1:
                        if self.filename.get() != '':
                            if os.path.isfile(self.filename.get()) == True:
                                try:
                                    draft = {'tweet':self.caption.get('1.0','end-1c'), 'media':self.filename.get()}
                                    json.dump(draft, f)
                                    f.write('\n')
                                    print('Draft saved!')
                                    mbox.showinfo('Success!','Draft has been saved')
                                except Exception as e:
                                    crashlog(exc = traceback.format_exc())
                                    exit()
                            else:
                                mbox.showinfo('Error','''No such file exists in this script's folder''')
                        else:
                            mbox.showinfo('Error','Please specify a media file')
                    else:
                        if self.caption.get('1.0','end-1c') != '':
                            draft = {'tweet':self.caption.get('1.0','end-1c')}
                            json.dump(draft, f)
                            f.write('\n')
                            print('Draft saved!')
                            mbox.showinfo('Success!','Draft has been saved')
                        else:
                            mbox.showinfo('Error','Your tweet is empty')

        def tweet():
            if cb1.get() == 1:
                if cb2.get() == 1:
                    if self.replyID.get() != '':
                        if self.filename.get() != '':
                            if len(self.caption.get('1.0','end-1c')) <= 280:
                                if os.path.isfile(self.filename.get()) == True:
                                    try:
                                        api.update_status_with_media(filename = self.filename.get(), status = self.caption.get('1.0','end-1c'), in_reply_to_status_id = self.replyID.get(), auto_populate_reply_metadata=True)
                                        print('Tweet sent!')
                                        os._exit(0)
                                    except Exception as e:
                                        crashlog(exc = traceback.format_exc())
                                        exit()
                                else:
                                    mbox.showinfo('Error','''No such file exists in this script's folder''')
                            else:
                                mbox.showinfo('Error',f'''Your tweet is over the character limit ({len(self.caption.get('1.0','end-1c'))})''')
                        else:
                            mbox.showinfo('Error','Please specify a media file')
                    else:
                        mbox.showinfo('Error','There is no tweet ID')
                else:
                    if self.caption.get('1.0','end-1c') != '':
                        if len(self.caption.get('1.0','end-1c')) <= 280:
                            if self.replyID.get() != '':
                                api.update_status(status = self.caption.get('1.0','end-1c'), in_reply_to_status_id = self.replyID.get(), auto_populate_reply_metadata=True)
                                print('Tweet sent!')
                                os._exit(0)
                            else:
                                mbox.showinfo('Error','There is no tweet ID')
                        else:
                            mbox.showinfo('Error',f'''Your tweet is over the character limit ({len(self.caption.get('1.0','end-1c'))})''')
                    else:
                        mbox.showinfo('Error','Your tweet is empty')

            else:
                if cb2.get() == 1:
                    if self.filename.get() != '':
                        if len(self.caption.get('1.0','end-1c')) <= 280:
                            if os.path.isfile(self.filename.get()) == True:
                                try:
                                    api.update_status_with_media(status = self.caption.get('1.0','end-1c'), filename = self.filename.get())
                                    print('Tweet sent!')
                                    os._exit(0)
                                except Exception as e:
                                    crashlog(exc = traceback.format_exc())
                                    exit()
                            else:
                                mbox.showinfo('Error','''No such file exists in this script's folder''')
                        else:
                            mbox.showinfo('Error',f'''Your tweet is over the character limit ({len(self.caption.get('1.0','end-1c'))})''')
                    else:
                        mbox.showinfo('Error','Please specify a media file')
                else:
                    if self.caption.get('1.0','end-1c') != '':
                        if len(self.caption.get('1.0','end-1c')) <= 280:
                            api.update_status(self.caption.get('1.0','end-1c'))
                            print('Tweet sent!')
                            os._exit(0)
                        else:
                            mbox.showinfo('Error',f'''Your tweet is over the character limit ({len(self.caption.get('1.0','end-1c'))})''')
                    else:
                        mbox.showinfo('Error','Your tweet is empty')

        self.sendTweet = tk.Button(top)
        self.sendTweet.place(relx=0.555, rely=0.540, height=85, width=200)
        self.sendTweet.configure(font="-size 14 -weight bold", command=tweet, text='Send Tweet!', pady="0", background=_bgcolor)

        self.saveDraft = tk.Button(top)
        self.saveDraft.place(relx=0.555, rely=0.800, height=45, width=98)
        self.saveDraft.configure(command=read_drafts, text='Drafts', pady="0", background=_bgcolor, font="-size 14 -weight bold")

        self.saveDraft = tk.Button(top)
        self.saveDraft.place(relx=0.732, rely=0.800, height=45, width=98)
        self.saveDraft.configure(font="-family {Segoe UI} -size 14 -weight bold", background=_bgcolor, pady="0", text='Save', command=save_drafts)

        if os.stat('temp.dump').st_size != 0:
            with open('temp.dump', 'r') as tempdump:
                templist = list(tempdump)
            open("temp.dump", "w").close()
            if str(templist[0]) != '':
                print('not empty')
                cdraft = json.loads(templist[0])
                BTDtweet = cdraft.get('tweet')
                self.caption.insert(1.0, BTDtweet)
                if 'TID' in cdraft:
                    self.Checkbutton1.select()
                    BTDTID = cdraft.get('TID')
                    self.replyID.delete(0,tk.END)
                    self.replyID.insert(0,BTDTID)
                if 'media' in cdraft:
                    self.Checkbutton2.select()
                    BTDmedia = cdraft.get('media')
                    self.filename.delete(0,tk.END)
                    self.filename.insert(0,BTDmedia)
        try: endis()
        except Exception as e:
            crashlog(exc = traceback.format_exc())
            exit()

if __name__ == '__main__':
    try: vp_start_gui()
    except Exception as e:
        crashlog(exc = traceback.format_exc())
        exit()
